src/aspire/source/xform.py

```diff
@@ -182,6 +182,48 @@ def _forward(self, im, indices):
         return im.filter(self.filter)
 
 
+class Add(Xform):
+    """
+    A Xform that add the density of a stack of 2D images (in the form of an Image object)
+    by offsetting a number.
+    """
+    def __init__(self, mean):
+        """
+        Initialize a Add Xform using a Numpy array of mean values.
+        :param mean: An ndarray of shape (n,)
+        """
+        super().__init__()
+        self.mean = mean
+
+    def _forward(self, im, indices):
+        return im + self.mean[indices]
+
+
+class FlipXform(Xform):
+    """
+    A `Xform` that applies phase flip to a stack of 2D images (as an Image object).
+    """
+    def __init__(self, filters):
+        """
+        Initialize the `FlipXform` using CTF `Filter` objects
+        :param filters: CTF filters of images
+        """
+        super().__init__()
+        self.filters = filters
+
+    def _forward(self, im, indices):
+        unique_filters = set(self.filters[indices])
+        im_in = im.asnumpy()
+        im_out = np.zeros_like(im_in)
+        for f in unique_filters:
+            idx_k = np.where(self.filters[indices] == f)[0]
+            flip_filter = LambdaFilter(f, np.sign)
+            if len(idx_k) > 0:
+                im_out[:, :, idx_k] = Image(im_in[:, :, idx_k]).filter(flip_filter).asnumpy()
```

Comment:
Does `Image` not support slicing already? If it does, it might be simpler to bypass the np arrays by just indexing the `Image` object, calling `filter`, and assigning this to an output `Image` object.

Reply:
I agree we should have been doing that.  However, the code that does the slicing in the Image class is totally wrong (it was obviously never used...).... It does work in the future... While it is a few lines to fix here, if we change it in this PR is will just be another merge conflict I get to resolve. 

```
class Image:
    def __init__(self, data):
        ensure(data.shape[0] == data.shape[1], 'Only square ndarrays are supported.')
        if data.ndim == 2:
            data = data[:, :, np.newaxis]

        self.data = data
        self.dtype = self.data.dtype
        self.shape = self.data.shape
        **self.n_images = self.shape[-1]**
        self.res = self.shape[0]

    def __getitem__(self, item):
        return **self.data[item]**

    def __setitem__(self, key, value):
        **self.data[key]** = value
```

Reply:
All right. Let's leave as is.

Reply:
I think in the original design we will let `Image` class deal with a subset data with the same property such as the filter from the  `ImageSource` class. I will leave as is.